### PR TITLE
add configuration files binding for --nv option

### LINKED ISSUE #2210: Missing EGL ICD when using --nv
### Version of Apptainer

`apptainer version 1.2.5`

### Expected behavior

When running with `--nv` apptainer needs to map EGL ICD `/usr/share/glvnd/egl_vendor.d/10_nvidia.json` same as NVIDIA Container Toolkit is doing per https://github.com/NVIDIA/nvidia-docker/issues/1520#issuecomment-1419071440
If not EGL initialisation fails (with `eglinfo`)

### Actual behavior

`/usr/share/glvnd/egl_vendor.d/10_nvidia.json` is not mapped and `eglinfo` fails

### Steps to reproduce this behavior

- install apptainer on a host with Nvidia GPU and proprietary drivers
- create a sif file `apptainer build --fakeroot eglinfo.sif eglinfo.recipe`

```
$ cat eglinfo.recipe
Bootstrap: docker
From: ubuntu:22.04
%post
    apt-get update -y
    apt-get install -y mesa-utils
    apt-get clean all
%runscript
    exec "${1+"$@"}"
```

- `apptainer run --nv eglinfo.sif eglinfo` (without the mapping)
```
[...]
Device platform:
eglinfo: eglInitialize failed
[...]
```
- `apptainer run --nv -B /usr/share/glvnd/egl_vendor.d/10_nvidia.json eglinfo.sif eglinfo` (with manually mapping the definition json)
```
[...]
Device platform:
EGL API version: 1.5
EGL vendor string: NVIDIA
EGL version string: 1.5
EGL client APIs: OpenGL_ES OpenGL
EGL extensions string:
[...]
```

### What OS/distro are you running

```
cat /etc/os-release
NAME="Red Hat Enterprise Linux"
VERSION="8.8 (Ootpa)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="8.8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Red Hat Enterprise Linux 8.8 (Ootpa)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:8::baseos"
HOME_URL="https://www.redhat.com/"
DOCUMENTATION_URL="https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 8"
REDHAT_BUGZILLA_PRODUCT_VERSION=8.8
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="8.8"
```

### How did you install Apptainer

from source

### PR DESCRIPTION
**Pull request by JasonYangShadow**
## Description of the Pull Request (PR):

add configuration files binding for --nv option

### This fixes or addresses the following GitHub issues:

 - Fixes #2210

#### Before submitting a PR, make sure you have done the following:

- [ ] Read the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [ ] Make sure all commits are signed-off with `git commit -s`, see the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md).
- [ ] Added changes to the [CHANGELOG](https://github.com/apptainer/apptainer/blob/main/CHANGELOG.md), if necessary according to the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md).
- [ ] Added tests to validate this PR, linted with `make check`  and tested this PR locally with a `make test`, and `make testall` if possible (see the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)).
- [ ] Based this PR against the appropriate branch according to the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md).
- [ ] Added myself as a contributor to the [Contributors File](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTORS.md).

### CODE REVIEW

**Comment by JasonYangShadow** [comment]
https://github.com/apptainer/apptainer/issues/2210#issuecomment-2421234163

File: CHANGELOG.md [hunk 1]
```diff
@@ -43,6 +43,7 @@ For older changes see the [archived Singularity change log](https://github.com/a
   post script instance.
 - Add `dnf` definition file bootstrap options as an alias to `yum`.
 - Add --sandbox option to apptainer pull.
+- Add configuration files binding for `--nv` option.
```
Comment (DrDaveD) [inline CHANGELOG.md:46]: Also have an entry about the additional files added to nvliblist.conf, e.g.:
```
- Add the confliguration files to nvliblist.conf that are recognized in the NVIDIA Container Toolkit, including files for EGL ICD.
```

File: etc/nvliblist.conf [hunk 2]
```diff
@@ -59,3 +59,12 @@ libOpenGL.so
...
 tls_test_.so
+
+# put non-library file paths here (must be absolute path)
+/usr/share/glvnd/egl_vendor.d/10_nvidia.json
+/usr/share/vulkan/icd.d/nvidia_icd.json
+/usr/share/vulkan/icd.d/nvidia_layers.json
+/usr/share/vulkan/implicit_layer.d/nvidia_layers.json
+/usr/share/egl/egl_external_platform.d/15_nvidia_gbm.json
+/usr/share/egl/egl_external_platform.d/10_nvidia_wayland.json
+/usr/share/nvidia/nvoptix.bin
```
Comment (DrDaveD) [inline etc/nvliblist.conf:70]: It looks like you lost the newline at the end of the file, please add that back.

File: internal/pkg/util/paths/resolve_test.go [hunk 3]
```diff
@@ -97,15 +97,30 @@ func TestSoLinks(t *testing.T) {
 
 func TestPaths(t *testing.T) {
 	// Very naive sanity test. Check we can find one lib and one binary without error
```
Comment (DrDaveD) [inline internal/pkg/util/paths/resolve_test.go:99]: Extend this comment to include a configuration file.

**Review by DrDaveD — CHANGES_REQUESTED** [review]
What error do you see if this tries to bind a configuration file during `build --nv` and the bindpoint doesn't exist?  I'd like to make sure it is sufficiently clear.

File: CHANGELOG.md (hunk 1 above)
Comment (JasonYangShadow) [reply CHANGELOG.md]: I revised the changelog entry for clarity.

File: etc/nvliblist.conf (hunk 2 above)
Comment (JasonYangShadow) [reply etc/nvliblist.conf]: Thanks! added

File: internal/pkg/util/paths/resolve_test.go (hunk 3 above)
Comment (JasonYangShadow) [reply internal/pkg/util/paths/resolve_test.go]: thanks! I refined the comment line.

**Comment by JasonYangShadow** [comment]
> What error do you see if this tries to bind a configuration file during `build --nv` and the bindpoint doesn't exist? I'd like to make sure it is sufficiently clear.

it'll show a warning message
```
[vagrant@localhost test]$ ap build -F --nv eglinfo.sif eglinfo.def 
INFO:    Starting build...
INFO:    Fetching OCI image...
28.2MiB / 28.2MiB [=======================================================================] 100 % 0.0 b/s 0s
INFO:    Extracting OCI image...
INFO:    Inserting Apptainer configuration...
INFO:    Running post scriptlet
WARNING: Could not find any nv files on this host!
+ apt-get update -y
....
```

**Review by DrDaveD — APPROVED** [review]

**Comment by DrDaveD** [comment]
> it'll show a warning message 
...
> WARNING: Could not find any nv files on this host!

But what happens if a file does exist on the host and there's no bind point for it in the container?

**Comment by JasonYangShadow** [comment]
> > it'll show a warning message
> > ...
> > WARNING: Could not find any nv files on this host!
> 
> But what happens if a file does exist on the host and there's no bind point for it in the container?

when I faked some of nvliblist.conf entries in the host (e.g., /usr/share/glvnd/egl_vendor.d/10_nvidia.json), just create an empty file in the host, the following warnings will be shown when do build
```
ap build -F --nv eglinfo.sif eglinfo.def 
INFO:    Starting build...
INFO:    Fetching OCI image...
28.2MiB / 28.2MiB [================================================================] 100 % 0.0 b/s 0s
INFO:    Extracting OCI image...
INFO:    Inserting Apptainer configuration...
INFO:    Running post scriptlet
WARNING: nv files may not be bound with --writable
WARNING: Skipping mount /usr/share/glvnd/egl_vendor.d/10_nvidia.json [files]: /usr/share/glvnd/egl_vendor.d/10_nvidia.json doesn't exist in container
```

**Comment by DrDaveD** [comment]
Ok that looks good.

**Pull request merged**